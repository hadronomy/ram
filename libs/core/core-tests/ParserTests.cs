namespace MemoryMachine.Tests;

using Antlr4.Runtime;
using MemoryMachine.Core.Parser;
using Snapshooter.Xunit;

/// <summary>
/// Tests for the memory machine assembly parser
/// </summary>
public class ParserTests
{


    [Fact(DisplayName = "Parse Program1 Should Create Valid AST")]
    public void ParseProgram1ShouldCreateValidAST()
    {
        // Arrange
        string input = Programs.Program1;

        var inputStream = new AntlrInputStream(input);
        var lexer = new MemoryMachineAssemblyLexer(inputStream);
        var tokens = new CommonTokenStream(lexer);
        var parser = new MemoryMachineAssemblyParser(tokens);

        var tree = parser.program();
        var stringRepresentation = tree.ToStringTree(parser);

        stringRepresentation.MatchSnapshot();
    }

    /// <summary>
    /// Parse Program2 Should Create Valid AST
    /// </summary>
    [Fact(DisplayName = "Parse Program2 Should  Create Valid AST")]
    public void ParseProgram2ShouldCreateValidAST()
    {
        // Arrange
        string input = Programs.Program2;

        var inputStream = new AntlrInputStream(input);
        var lexer = new MemoryMachineAssemblyLexer(inputStream);
        var tokens = new CommonTokenStream(lexer);
        var parser = new MemoryMachineAssemblyParser(tokens);

        var tree = parser.program();
        var stringRepresentation = tree.ToStringTree(parser);

        stringRepresentation.MatchSnapshot();
    }

    [Fact(DisplayName = "Parse Program3 Should Create Valid AST")]
    public void ParseProgram3ShouldCreateValidAST()
    {

        string input = Programs.Program3;

        var inputStream = new AntlrInputStream(input);
        var lexer = new MemoryMachineAssemblyLexer(inputStream);
        var tokens = new CommonTokenStream(lexer);
        var parser = new MemoryMachineAssemblyParser(tokens);

        var tree = parser.program();
        var stringRepresentation = tree.ToStringTree(parser);

        stringRepresentation.MatchSnapshot();
    }

    [Fact(DisplayName = "Parse Program4 Should Create Valid AST")]
    public void ParseProgram4ShouldCreateValidAST()
    {
        string input = Programs.Program4;

        var inputStream = new AntlrInputStream(input);
        var lexer = new MemoryMachineAssemblyLexer(inputStream);
        var tokens = new CommonTokenStream(lexer);
        var parser = new MemoryMachineAssemblyParser(tokens);

        var tree = parser.program();
        var stringRepresentation = tree.ToStringTree(parser);

        stringRepresentation.MatchSnapshot();
    }

    [Fact(DisplayName = "Parse Program5 Should Create Valid AST")]
    public void ParseProgram5ShouldCreateValidAST()
    {
        string input = Programs.Program5;

        var inputStream = new AntlrInputStream(input);
        var lexer = new MemoryMachineAssemblyLexer(inputStream);
        var tokens = new CommonTokenStream(lexer);
        var parser = new MemoryMachineAssemblyParser(tokens);

        var tree = parser.program();
        var stringRepresentation = tree.ToStringTree(parser);

        stringRepresentation.MatchSnapshot();
    }

    [Fact(DisplayName = "Parse Program6 Should Create Valid AST")]
    public void ParseProgram6ShouldCreateValidAST()
    {
        string input = Programs.Program6;

        var inputStream = new AntlrInputStream(input);
        var lexer = new MemoryMachineAssemblyLexer(inputStream);
        var tokens = new CommonTokenStream(lexer);
        var parser = new MemoryMachineAssemblyParser(tokens);

        var tree = parser.program();
        var stringRepresentation = tree.ToStringTree(parser);

        stringRepresentation.MatchSnapshot();
    }

    [Fact(DisplayName = "Parse Program7 Should Create Valid AST")]
    public void ParseProgram7ShouldCreateValidAST()
    {
        string input = Programs.Program7;

        var inputStream = new AntlrInputStream(input);
        var lexer = new MemoryMachineAssemblyLexer(inputStream);
        var tokens = new CommonTokenStream(lexer);
        var parser = new MemoryMachineAssemblyParser(tokens);

        var tree = parser.program();
        var stringRepresentation = tree.ToStringTree(parser);

        stringRepresentation.MatchSnapshot();
    }

    [Fact(DisplayName = "Parse InsertionSort Should Create Valid AST")]
    public void ParseArrayToRegistersShouldCreateValidAST()
    {
        string input = Programs.ArrayToRegisters;

        var inputStream = new AntlrInputStream(input);
        var lexer = new MemoryMachineAssemblyLexer(inputStream);
        var tokens = new CommonTokenStream(lexer);
        var parser = new MemoryMachineAssemblyParser(tokens);

        var tree = parser.program();
        var stringRepresentation = tree.ToStringTree(parser);

        stringRepresentation.MatchSnapshot();
    }
}
