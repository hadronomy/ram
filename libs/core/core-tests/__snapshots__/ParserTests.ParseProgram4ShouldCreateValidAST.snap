(program (line (comment # Programa que lee los elementos la cinta de entrada hasta)) \n (line (comment # que encuentre un 0 y escribe el doble del valor leído a la)) \n (line (comment # cinta de salida (el valor 0 no se copia))) \n\n (line (label_definition bucle : (instruction READ (argument (direct 1))))) \n (line (instruction LOAD (argument (direct 1)))) \n (line (instruction JZERO (argument (label fin)))) \n (line (instruction LOAD (argument (direct 1)))) \n (line (instruction MUL (argument (immediate = 2)))) \n (line (instruction STORE (argument (direct 1)))) \n (line (instruction WRITE (argument (direct 1)))) \n (line (instruction JUMP (argument (label bucle)))) \n (line (label_definition fin : (instruction HALT))) \n <EOF>)
